### PR TITLE
[WIP] Adding Restore3 command

### PR DESCRIPTION
**Pull request by emgarten**
This change adds the NuGet.Build.Tasks package containing NuGet's MSBuild task for restore. The Restore3 command will call the Restore target in NuGet.targets and convert the command line parameters into MSBuild properties.

### CODE REVIEW

**Comment by emgarten** [comment]
//cc @rrelyea

**Comment by jasonwilliams200OK** [comment]
> This branch has conflicts that must be resolved

**Comment by emgarten** [comment]
//cc @eerhardt @piotrpMSFT @rohit21agrawal

**Comment by rohit21agrawal** [comment]
LGTM. :shipit:

File: src/dotnet/project.json [hunk 1]
```diff
@@ -26,6 +26,8 @@
     }
   },
   "dependencies": {
+    "NuGet.CommandLine.XPlat": "3.6.0-beta.1.msbuild.1",
```
Comment (eerhardt) [inline src/dotnet/project.json]: I don't think these references are needed anymore right?  We never actually load these assemblies into our process.  Instead we invoke MSBuild /t:Restore in a separate process.

File: src/dotnet/commands/dotnet-restore3/Program.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace Microsoft.DotNet.Tools.Restore3
+{
+    public partial class Restore3Command
+    {
+        public static int Run(string[] args)
+        {
+            DebugHelper.HandleDebugSwitch(ref args);
+
+            return Restore3CommandApp.Execute(args);
```
Comment (eerhardt) [inline src/dotnet/commands/dotnet-restore3/Program.cs]: Typically, the arg definitions happen in this Program.cs file, and the "logic" happens in the other file.  See dotnet-publish or dotnet-run for an example.

That being said, there really isn't "logic" in this command.  What do you think about collapsing these two files into a single file?  I think it would be easier to read.  We can spit them later, if more logic gets added.

File: src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            CommandLineApplication cmd = new CommandLineApplication
+            {
+                Name = "restore3",
+                FullName = "restore3",
+                Description = "restore for msbuild"
+            };
+
+            cmd.HelpOption("-h|--help");
+
+            var argRoot = cmd.Argument(
```
Comment (eerhardt) [inline src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs]: Do we want this, or do we want to support 'pass through' args?  like "dotnet run" does.  Basically allow the user to pass any arg through to the MSBuild process.

To allow this, you need to set `throwOnUnexpectedArg` = false, and grab the RemainingArguments from the CommandLineApplication.

File: src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                    "[root]",
+                    "Optional path to a project file or MSBuild arguments.",
+                    multipleValues: true);
+
+            var sourceOption = cmd.Option(
```
Comment (eerhardt) [inline src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs]: I don't see `-f|--fallbacksource <FEED>` as an option.  Is that still supported?

Comment (eerhardt) [inline src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs]: I also don't see the `-v|--verbosity` option.  Are we expecting to use msbuild's verbosity?

File: src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+                if (ignoreFailedSourcesOption.HasValue())
+                {
+                    msbuildArgs.Add($"/p:RestoreIgnoreFailedSources=true");
+                }
+
+                // Add any additional values passed
+                msbuildArgs.AddRange(argRoot.Values);
+
+                return new MSBuildForwardingApp(msbuildArgs.ToArray()).Execute();
```
Comment (eerhardt) [inline src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs]: There's no need that MSBuildForwardingApp strictly needs a `string[]`.  We should change that class to take an IEnumerable<string> instead, and then get rid of the ToArray() call here.

**Comment by eerhardt** [comment]
:clock1:

File: src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs (hunk 4 above)
Comment (emgarten) [inline src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs]: FallbackSource has been removed, I know it's used some places in the CLI build still. Is there a good reason to keep this instead of just having the user specify all sources with ``-s``?

``--verbosity`` ended up being a converter from the nuget level names to the msbuild level names, but going forward it seemed really unnecessary. Going forward I think this command should use whatever verbosity setting the rest of the commands use, and for now ``/v:`` lets you set it directly on msbuild, which controls the nuget restore output.

File: src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs (hunk 3 above)
Comment (emgarten) [inline src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs]: That works today because the msbuild args start with ``/`` and this parser looks for ``-``. Unless you can also pass ``-`` to msbuild, is that possible?

I don't see a way to validate unexpected args in msbuild, so if we set this to false there wouldn't be any help.

File: src/dotnet/commands/dotnet-restore3/Program.cs (hunk 2 above)
Comment (emgarten) [inline src/dotnet/commands/dotnet-restore3/Program.cs]: I can do that, I was just trying to follow the pattern of the other commands.

File: src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs (hunk 4 above)
Comment (eerhardt) [inline src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs]: > FallbackSource has been removed, I know it's used some places in the CLI build still. Is there a good reason to keep this instead of just having the user specify all sources with -s?

According to https://github.com/NuGet/Home/issues/2477, `-s` will completely override what is in NuGet.config.  This seems limiting since in a lot of cases you only want to add a source for this specific invocation.  Not completely overriding NuGet.config.  But there are other cases where you DO want to override.

Will both scenarios be supported by `-s`?

File: src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs (hunk 3 above)
Comment (eerhardt) [inline src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs]: `dotnet build3 -v:diag` works today.  MSBuild knows how to look for `/` and for `-`.

I would assume MSBuild would validate the unexpected args.

File: src/dotnet/project.json (hunk 1 above)
Comment (emgarten) [inline src/dotnet/project.json]: The non-msbuild based restore command still exists to allow restoring xproj, this reference is needed for that.

Comment (eerhardt) [inline src/dotnet/project.json]: @piotrpMSFT just removed this reference and changed the non-msbuild based restore command to call NuGet in a separate process.  See https://github.com/dotnet/cli/pull/4024

**Comment by emgarten** [comment]
Updated

File: src/dotnet/project.json (hunk 1 above)
Comment (emgarten) [inline src/dotnet/project.json]: I'll remove them and see what is needed.

File: src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs (hunk 4 above)
Comment (emgarten) [inline src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs]: ``-s`` just overrides, I think we need to look at fallback source, that scenario can be useful although it is rare, and at minimum it needs a new name to help with #2477. I'm not willing to add it back under the fallback name here since this is a new command and it should stay as clean as possible.

File: src/dotnet/project.json (hunk 1 above)
Comment (eerhardt) [inline src/dotnet/project.json]: Were you able to remove these?  I still can't see where they are necessary.

---
In reply to: [74501423](https://github.com/dotnet/cli/pull/4035#discussion_r74501423) [](ancestors = 74501423)

File: src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs (hunk 3 above)
Comment (emgarten) [inline src/dotnet/commands/dotnet-restore3/Restore3CommandApp.cs]: Sounds good, I'll set ``throwOnUnexpectedArg`` to false

**Comment by eerhardt** [comment]
:shipit:  once addressing the comment about needing the NuGet dependencies on src\dotnet\project.json

**Comment by emgarten** [comment]
@eerhardt removing the dependencies from *src\dotnet\project.json* appears to cause several problems, the targets file doesn't get deployed and there is a compile error on a reference to a NuGet assembly. The dotnet project itself restores and builds fine, but there must be some *nearest package* changes that end up bringing in the wrong bits.

File: build_projects/dotnet-cli-build/project.json [hunk 6]
```diff
@@ -20,7 +20,8 @@
     "System.Runtime.Serialization.Primitives": "4.1.1",
     "System.Xml.XmlSerializer": "4.0.11",
     "WindowsAzure.Storage": "6.2.2-preview",
-    "NuGet.CommandLine.XPlat": "3.5.0-rc1-1697",
+    "NuGet.Build.Tasks": "3.6.0-beta.1.msbuild.1",
```
Comment (eerhardt) [inline build_projects/dotnet-cli-build/project.json]: Our private build tasks project should need to reference `NuGet.Build.Tasks`.  Why is this needed?

**Comment by eerhardt** [comment]
> the targets file doesn't get deployed

Ah, right. I overlooked that part.  I think it would be better if the `NuGet.Build.Tasks` dependency was brought in by either https://github.com/dotnet/cli/blob/feature/msbuild/src/redist/project.json or https://github.com/dotnet/cli/blob/feature/msbuild/src/tool_nuget/project.json.  My thought is the 1st one.

/cc @piotrpMSFT, who added these new projects to get his opinion on which would be better.

**Comment by eerhardt** [comment]
> there is a compile error on a reference to a NuGet assembly. The dotnet project itself restores and builds fine, but there must be some nearest package changes that end up bringing in the wrong bits.

I think that's because a new NuGet dependency was added in the latest bits that you haven't updated:

See https://github.com/dotnet/cli/blob/feature/msbuild/src/tool_nuget/project.json#L11

**Comment by emgarten** [comment]
>  I think it would be better if the NuGet.Build.Tasks dependency was brought in by either https://github.com/dotnet/cli/blob/feature/msbuild/src/redist/project.json 

I'll try putting it there

> I think that's because a new NuGet dependency was added in the latest bits that you haven't updated:

That could be the problem, not sure how find and replace wasn't catching that one

**Comment by emgarten** [comment]
Thanks for the pointers @eerhardt updating and moving the references solve the issue. I've checked this in, if the build task needs to go under another project I can move it.
